README
NAME
    App::MiseEnPlace - A place for everything and everything in its place

VERSION
    version 0.14

SYNOPSIS
    See 'pod mise' for usage details.

AUTHOR
    John SJ Anderson <genehack@genehack.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by John SJ Anderson.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Revision History for App-MiseEnPlace
====================================

0.14
----

commit 8e556929f5a821aed6e419ef8e65e6bc7d0e3151
Author: John SJ Anderson <genehack@genehack.org>
Date:   Sat Nov 13 15:51:32 2010 -0500

    Add 'remove-bin-links' option

commit 93a1dbeeac9d63d2b6a7eb7a7fae56f81b46bb7e
Author: John SJ Anderson <genehack@genehack.org>
Date:   Sat Nov 13 15:28:06 2010 -0500

    output improvements; check existing link targets
    
    * colored output
    * piped thru pager in semi-intelligent fashion
    * better output when running non-verbose
    
    * if target is a link that points to something other
      than the configured source, remove it and make the
      correct new link.

commit 581092df2764b170874cba2ca8f0c29bfe8dab5b
Author: John SJ Anderson <genehack@genehack.org>
Date:   Fri Nov 12 00:12:46 2010 -0500

    released 0.13

v0.13 2010-11-12
----------------

commit a5475f101c240ec1a6bb798d3e5947e7e95b4b3a
Author: John SJ Anderson <genehack@genehack.org>
Date:   Fri Nov 12 00:00:32 2010 -0500

    specify File::Path version to lock in new API

commit 303283afffc94eb70a748d2800ba7430345541d9
Author: John SJ Anderson <genehack@genehack.org>
Date:   Mon Nov 8 20:56:12 2010 -0500

    released 0.12

v0.12 2010-11-08
----------------

commit 9845b6b2db61b1123c60fc4e327f9763a4eaef2c
Author: John SJ Anderson <genehack@genehack.org>
Date:   Mon Nov 8 20:51:19 2010 -0500

    fix default config file location

commit fe1f712b65d18bd16c813176518e4b701a69684a
Author: John SJ Anderson <genehack@genehack.org>
Date:   Mon Nov 8 20:50:58 2010 -0500

    add support for DIR keyword in create/links

commit 69b9ba8f17ce8fdda4e35c86e000389734e69d39
Author: John SJ Anderson <genehack@genehack.org>
Date:   Mon Nov 8 20:41:02 2010 -0500

    whitespace

commit b3e04ffc02671b9737b803d9abef1df695a7cb9f
Author: John SJ Anderson <genehack@genehack.org>
Date:   Mon Nov 8 20:40:52 2010 -0500

    handle missing config a bit more gracefully

commit ba405c37c22b5468e7ae5feb5bfbcb996964728e
Author: John SJ Anderson <genehack@genehack.org>
Date:   Sun Nov 7 23:32:08 2010 -0500

    released

===================================
End of changes in the last 365 days
===================================


lib/App/MiseEnPlace.pm
package App::MiseEnPlace;
BEGIN {
  $App::MiseEnPlace::VERSION = '0.14';
}
BEGIN {
  $App::MiseEnPlace::AUTHORITY = 'cpan:GENEHACK';
}
# ABSTRACT: A place for everything and everything in its place


use strict;
use warnings;
use 5.010;

use base 'App::Cmd::Simple';
use autodie;
use Carp;
use File::Basename;
use File::Path 2.08  qw/ make_path /;
use File::Path::Expand;
use Mouse;
use Term::ANSIColor;
use Try::Tiny;
use YAML        qw/ LoadFile /;

has 'bindir' => (
  is => 'rw' ,
  isa => 'Str' ,
  required => 1 ,
  default => sub { expand_filename '~/bin/' } ,
  lazy => 1 ,
);

has 'config_file' => (
  is       => 'rw' ,
  isa      => 'Str' ,
  default  => "$ENV{HOME}/.mise" ,
  lazy     => 1 ,
  required => 1 ,
);

has 'directories' => (
  is  => 'rw' ,
  isa => 'ArrayRef[Str]' ,
);

has 'homedir' => (
  is       => 'rw' ,
  isa      => 'Str' ,
  required => 1 ,
  lazy     => 1 ,
  default  => $ENV{HOME} ,
);

has 'links' => (
  is  => 'rw' ,
  isa => 'ArrayRef[ArrayRef[Str]]' ,
);

has 'verbose' => (
  is      => 'rw' ,
  isa     => 'Bool' ,
  default => 0 ,
);

sub opt_spec {
  return (
    [ 'config|C=s' => 'config file location (default = ~/.mise)' ] ,
    [ 'remove-bin-links|R' => 'remove all links from ~/bin at beginning of run' ] ,
    [ 'verbose|v' => 'be verbose' ] ,
    [ 'version|V' => 'show version' ] ,
  );
}

sub validate_args {
  my( $self , $opt , $args ) = @_;

  $self->usage_error( "No args needed" ) if @$args;

  if ( $opt->{version} ) {
    say $App::MiseEnPlace::VERSION;
    exit;
  }

  $self->config_file( $opt->{config} ) if $opt->{config};
  $self->verbose( $opt->{verbose} ) if $opt->{verbose};

}

sub execute {
  my( $self , $opt , $args ) = @_;

  # set up colored output if we page thru less
  # also exit pager immediately if <1 page of output
  $ENV{LESS} = 'RF';

  # don't catch any errors here; if this fails we just output stuff like
  # normal and nobody is the wiser.
  eval 'use IO::Page';

  $self->_load_configs;

  $self->_create_dir( $_ )  for ( @{ $self->directories } );

  if ( $opt->{remove_bin_links} and -e -d $self->bindir ) {
    my $bin = $self->bindir;
    opendir( my $dh , $bin );
    while ( readdir $dh ) {
      next unless -l "$bin/$_";
      unlink "$bin/$_";
      say colored('UNLINK' , 'bright_red' ) ,
        " ~/bin/$_" if $opt->{verbose};
    }
    closedir( $dh );
  }

  $self->_create_link( $_ ) for ( @{ $self->links } );

}

sub _create_dir {
  my( $self , $dir ) = @_;

  my $msg;

  given( $dir ) {
    when( -e -d ) {
      $msg = colored('exists ','green') if $self->verbose;
    }
    when( -e and ! -l ) {
      $msg = colored('ERROR: blocked by non-dirctory','bold white on_red');
    }
    default {
      make_path $dir;
      $msg = colored('created','bold black on_green');
    }
  }

  my $home = $self->homedir;
  if ( $msg ) {
    $dir =~ s/^$home/~/;
    say "[ DIR] $msg $dir";
  }
}

sub _create_link {
  my( $self , $linkpair ) = @_;

  my( $src , $target ) = @$linkpair;

  my $msg;

  if ( ! -e $src ) {
    $msg = colored( 'ERROR:  src does not exist' , 'bold white on_red' )
  }
  elsif( -e -l $target ) {
    if ( readlink $target eq $src ) {
      $msg = colored('exists ','green') if $self->verbose;
    }
    else {
      unlink $target;
      symlink $src , $target;
      $msg = colored( 'fixed' , 'bold black on_bright_yellow' ) . '  ';
    }
  }
  elsif ( -e $target ) {
    $msg = colored( 'ERROR:  blocked by existing file' , 'bold white on_red' );
  }
  else {
    symlink $src , $target;
    $msg = colored( 'created' , 'bold black on_green' );
  }
  my $home = $self->homedir;

  if ( $msg ) {
    $src    =~ s/^$home/~/;
    $target =~ s/^$home/~/;
    say "[LINK] $msg $src -> $target";
  }
}

sub _load_configs {
  my( $self ) = shift;

  unless ( -e $self->config_file ) {
    say "Whoops, it looks like you don't have a ~/.mise file yet.";
    say "Please review the documentation, create one, and try again.";
    exit;
  }

  my $base_config = _load_config_file( $self->config_file );

  my @links = map { _parse_linkpair( $_ , $self->homedir) }
    @{ $base_config->{create}{links} };

  my @dirs = map { _prepend_dir( $_ , $self->homedir) }
    @{ $base_config->{create}{directories} };

  my @managed_dirs = map { glob _prepend_dir( $_ , $self->homedir ) }
    @{ $base_config->{manage} };

  for my $managed_dir ( @managed_dirs ) {
    my $mise_file = "$managed_dir/.mise";
    if ( -e -r $mise_file ) {
      my $config = _load_config_file( $mise_file );

      for ( @{ $config->{create}{directories} } ) {
        push @dirs , _prepend_dir( $_ , $managed_dir );
      }

      for ( @{ $config->{create}{links} } ) {
        push @links , _parse_linkpair( $_ , $managed_dir );
      }
    }
  }

  $self->directories( \@dirs );

  $self->links( $self->_parse_create_links( \@links ) );
}

sub _load_config_file {
  my $file = shift;

  my $config;

  try { $config = LoadFile( glob($file) ) }
  catch {
    say "Failed to parse config file $file:\n\t$_";
    exit;
  };

  return $config;

}

sub _parse_create_links {
  my( $self, $link_array ) = @_;

  my( %link_targets , @links );

  for my $link_pair ( @$link_array ) {
    my( $src , $target ) = ( %$link_pair );

    my $src_base = basename( $src );

    $target = $self->bindir if $target =~ m'BIN$';
    $target = "$target$src_base" if $target =~ m|/$|;

    if (exists $link_targets{$target} ) {
      say "ERROR: Attempting to create multiple links to the same target:";
      printf "%s -> %s\n%s -> %s\n" ,
        $link_targets{$target} , $target , $src , $target;
    }

    $link_targets{$target} = $src;

    push @links , [ $src , $target ];
  }

  return \@links;

}

sub _parse_linkpair {
  confess "BAD ARGS" unless
    my( $linkpair , $dir ) = @_;

  confess "BAD LINKPAIR" unless
    my( $src , $target ) = ( %$linkpair );

  # this lets 'DIR' turn into enclosing directory
  $src = '' if $src eq 'DIR';

  $src    = _prepend_dir( $src , $dir );
  $target = _prepend_dir( $target , $dir ) unless $target eq 'BIN';

  return { $src => $target };
}

sub _prepend_dir {
  confess "BAD ARGS" unless
    my( $base , $dir ) = @_;

  return expand_filename $base if $base =~ m|^~|;
  return "$dir/$base" unless $base =~ m|^/|;
  return $base;
}

1;

__END__
=pod

=head1 NAME

App::MiseEnPlace - A place for everything and everything in its place

=head1 VERSION

version 0.14

=head1 SYNOPSIS

See 'pod mise' for usage details.

=head1 AUTHOR

John SJ Anderson <genehack@genehack.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by John SJ Anderson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

BEGIN { require 5.010; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'A place for everything and everything in its place',
  'AUTHOR' => 'John SJ Anderson <genehack@genehack.org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'App-MiseEnPlace',
  'EXE_FILES' => [
    'bin/mise'
  ],
  'LICENSE' => 'perl',
  'NAME' => 'App::MiseEnPlace',
  'PREREQ_PM' => {
    'App::Cmd::Simple' => '0',
    'Carp' => '0',
    'File::Basename' => '0',
    'File::Path' => '2.08',
    'File::Path::Expand' => '0',
    'FindBin' => '0',
    'Mouse' => '0',
    'Term::ANSIColor' => '0',
    'Try::Tiny' => '0',
    'YAML' => '0',
    'autodie' => '0'
  },
  'VERSION' => '0.14',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



